types/types.go

```diff
@@ -5,6 +5,46 @@ import (
 	"time"
 )
 
+// ImageTransport is a top-level namespace for ways to refer to an image.
+// It should generally correspond to ImageSource/ImageDestination implementations.
+//
+// Note that ImageTransport is based on "ways the users refer to images", not necessarily on the underlying physical transport.
+// For example, all Docker References would be used within a single "docker" transport, regardless of whether the images are pulled over HTTP or HTTPS
+// (or, even, IPv4 or IPv6).
+//
+// OTOH all images using the same transport should (apart from format versions), be interoperable.
+// For example, several different ImageTransport implementations may be based on local filesystem paths,
+// but using completely different formats for the contents of that path (a single tar file, a directory containing tarballs, a fully expanded container filesystem, ...)
+// See also containers/image.KnownTransports.
+type ImageTransport interface {
+	// Name returns the name of the transport, which must be unique among other transports.
+	Name() string
+	// ParseReference converts a string, which should not start with the ImageTransport.Name prefix, into an ImageReference.
+	ParseReference(reference string) (ImageReference, error)
+}
+
+// ImageReference is an abstracted way to refer to an image or a namespace containing images, namespaced within an ImageTransport
+type ImageReference interface {
+	Transport() ImageTransport
+	// ConvenientString returns a string representation convenient for the user, even if perhaps syntactically a bit ambiguous.
+	// For example, for Docker, this may exclude the default Docker Hub hostname or the "library" namespace.
+	// This is not purely for display: parsing the string into an ImageReference must result in an equivalent reference.
+	// It is perfectly acceptable for ConvenientString() and FullyExplicitString() to return the same value if the underlying references do not have format variants.
+	ConvenientString() string
+	// FullyExplicitString returns a string representation which is fully explicit about its semanitcs, even if a bit inconvenient to use.
+	// For example, for Docker, this should include the default Docker Hub hostname or the "library" namespace.
+	// This is not purely for display: parsing the string into an ImageReference must result in an equivalent reference.
+	// It is highly recommended for FullyExplicitString to be designed so that all parent namespaces of a reference are initial prefixes of FullyExplicitString().
+	// It is perfectly acceptable for ConvenientString() and FullyExplicitString() to return the same value if the underlying references do not have format variants.
+	FullyExplicitString() (string, error)
```

Comment:
`Image{,Source,Destination}` would then, instead of providing `IntendedDockerReference` and `CanonicalDockerReference`, simply provide a `Reference` returning what they were created with, and the mapping intelligence would all be in the `ImageReference` implementation.